compiler/rustc_trait_selection/src/solve/search_graph/mod.rs

```diff
@@ -179,4 +180,26 @@ impl<'tcx> SearchGraph<'tcx> {
             true
         }
     }
+
+    pub(super) fn with_new_goal(
+        &mut self,
+        tcx: TyCtxt<'tcx>,
+        canonical_goal: CanonicalGoal<'tcx>,
+        loop_body: impl FnMut(&mut Self) -> Option<QueryResult<'tcx>>,
+    ) -> QueryResult<'tcx> {
+        match self.try_push_stack(tcx, canonical_goal) {
+            Ok(()) => {}
+            // Our goal is already on the stack, eager return.
+            Err(response) => return response,
+        }
+
+        // We may have to repeatedly recompute the goal in case of coinductive cycles,
+        // check out the `cache` module for more information.
+        let result =
+            self.repeat_while_none(|this| this.deal_with_overflow(tcx, canonical_goal), loop_body);
+
+        self.try_finalize_goal(tcx, canonical_goal, result);
```

Comment:
this changes behavior. we currently never repeat here, the check in `try_finalize_goal` has to happen inside of `repeat_while_none`.